src/System.IO.Compression/src/System/IO/Compression/InflaterZlib.cs

```diff
@@ -0,0 +1,275 @@
+        private bool _isDisposed;                           // Prevents multiple disposals
+        private ZLibNative.ZLibStreamHandle _zlibStream;    // The handle to the primary underlying zlib stream
+        private GCHandle? _inputBufferHandle;               // The handle to the buffer that provides input to _zlibStream
+        private readonly object _syncLock = new object();   // Used to make writing to unmanaged structures atomic 
+
+        #region Exposed Members
+
+        /// <summary>
+        /// Initialized the Inflater with a windowBits size set to -15 to denote
+        /// raw stream inflation i.e. header/footers are read as regular data
+        /// </summary>
+        internal InflaterZlib()
+        {
+            _finished = false;
+            _isDisposed = false;
+            InflateInit(ZLibNative.Deflate_DefaultWindowBits);
+        }
+
+        public int AvailableOutput
+        {
+            get
+            {
+                return (int)_zlibStream.AvailOut;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if the end of the stream has been reached.
+        /// </summary>
+        public bool Finished()
+        {
+            return _finished && _zlibStream.AvailIn == 0 && _zlibStream.AvailOut == 0;
+        }
+
+        public int Inflate(byte[] bytes, int offset, int length)
+        {
+            Debug.Assert(null != bytes, "Can't pass in a null output buffer!");
+
+            // If Inflate is called on an invalid or unready inflater, return 0 to indicate no bytes
+            // have been read
+            if (((IInflater)this).NeedsInput() || _inputBufferHandle == null || _inputBufferHandle.Value == null || length == 0)
+                return 0;
+
+            // State is valid; attempt inflation
+            try
+            {
+                int bytesRead;
+                ZLibNative.ErrorCode errc = ReadInflateOutput(bytes, offset, length, ZLibNative.FlushCode.NoFlush, out bytesRead);
+                if (errc == ZLibNative.ErrorCode.StreamEnd)
+                    _finished = true;
+                return bytesRead;
+            }
+            finally
+            {
+                // Before returning, make sure to release input buffer if necesary:
+                if (0 == _zlibStream.AvailIn && _inputBufferHandle.HasValue)
+                    DeallocateInputBufferHandle();
+            }
+        }
+
+        bool IInflater.NeedsInput()
+        {
+            return _zlibStream.AvailIn == 0;
+        }
+
+        public void SetFileFormatReader(IFileFormatReader reader)
+        {
+            // Rather than reading raw data and using a FormatReader to interpret
+            // headers/footers manually, we instead reset the zlib stream to parse
+            // that information for us.
+            ZLibNative.ErrorCode error;
+            try
+            {
+                error = _zlibStream.InflateReset(reader.ZLibWindowSize);
```

Comment:
Yes, if Zlib doesn't support reading the header then subsequent inflation calls will fail due to a Z_DATA_ERROR. This method actually sets the WindowBits size which happens to also enable setting of header/footer parsing, so its a bit of a misnomer (although still does what the method promises in the case where reader if for GZip).

I wrote it such that if ZLib supports the header with the given window size as it's designation, then we do as well. Since GZip is the only supported option for this though, I will add a Debug.Assert for it.